Public/Src/Tools/Execution.Analyzer/Analyzers.Partner/ContentPlacementAnalyzer.cs

```diff
@@ -0,0 +1,532 @@
+                m_outputWriter.WriteEndObject();
+            }
+            m_outputWriter.WriteEndArray();
+            m_outputWriter.WritePropertyName("OutputPips");
+            m_outputWriter.WriteStartArray();
+            for (int i = 0; i < artifact.NumOutputPips; ++i)
+            {
+                m_outputWriter.WriteStartObject();
+                WriteJsonPropertyToStream("SemiStableHash", artifact.OutputPips[i].SemiStableHash);
+                WriteJsonPropertyToStream("Priority", artifact.OutputPips[i].Priority);
+                WriteJsonPropertyToStream("Weight", artifact.OutputPips[i].Weight);
+                WriteJsonPropertyToStream("TagCount", artifact.OutputPips[i].TagCount);
+                WriteJsonPropertyToStream("DependencyCount", artifact.OutputPips[i].DependencyCount);
+                WriteJsonPropertyToStream("InputCount", artifact.OutputPips[i].InputCount);
+                WriteJsonPropertyToStream("OutputCount", artifact.OutputPips[i].OutputCount);
+                WriteJsonPropertyToStream("SemaphoreCount", artifact.OutputPips[i].SemaphoreCount);
+                WriteJsonPropertyToStream("StartTimeTicks", artifact.OutputPips[i].StartTimeTicks);
+                WriteJsonPropertyToStream("Type", artifact.OutputPips[i].Type);
+                WriteJsonPropertyToStream("ExecutionLevel", artifact.OutputPips[i].ExecutionLevel);
+                m_outputWriter.WriteEndObject();
+            }
+            m_outputWriter.WriteEndArray();
+            m_outputWriter.WriteEndObject();
+            // done...
+        }
+
+        private void WriteJsonPropertyToStream<T>(string name, T value)
+        {
+            m_outputWriter.WritePropertyName(name);
+            m_outputWriter.WriteValue(value);
+        }
+
+        public bool IncludePip(Pip pip)
+        {
+            return pip.PipType != PipType.HashSourceFile;
+        }
+
+        public override void Prepare()
+        {
+            string outputFile = $"{Path.Combine(Input.CachedGraphDirectory, m_relativeOutputFile)}";
+            Console.WriteLine($"{GetType().FullName} starting. Output will be written to [{outputFile}]");
+            m_outputWriter = new JsonTextWriter(new StreamWriter(outputFile));
+            Contract.Requires(File.Exists(outputFile), $"Could not create [{outputFile}]...");
+        }
+
+        public override void Dispose()
+        {
+            if (m_outputWriter != null)
+            {
+                m_outputWriter.Close();
+            }
+            base.Dispose();
+        }
+
+        public override void FileArtifactContentDecided(FileArtifactContentDecidedEventData data)
+        {
+            string extension = Path.GetExtension(data.FileArtifact.Path.ToString(CachedGraph.Context.PathTable));
+            AddToDictionaryWithListAsValue(m_artifactsPerExtension, extension, data.FileArtifact);
+            CountInDictionary(m_countFilesPerExtension, extension);

```

Comment:
Technically this one (and perhaps the other use cases of CountInDictionary) can be replaced with a single walk of the items i.e.

```
m_countFilesPerExtension = m_artifactsPerExetnsion
    .Select(kv => new KeyValuePair<string,int>(kv.Key, kv.Value.Count))
    .ToDictionary()
```

Given the ratio of extensions vs file artifacts I suspect this would be a bit more expedient.

Reply:
note: no need to fix, leaving this as your call. Merely sharing my thoughts. 

---
In reply to: [314072670](https://github.com/microsoft/BuildXL/pull/734#discussion_r314072670) [](ancestors = 314072670)